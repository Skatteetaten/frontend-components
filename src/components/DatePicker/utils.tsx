export const DEFAULTFORMATDATE = (date: Date | null | undefined): string => {
  return date
    ? Intl.DateTimeFormat('no', {
        year: 'numeric',
        month: '2-digit',
        day: 'numeric',
      }).format(date)
    : '';
};

export const DEFAULTPARSEDATEFROMSTRING = (date: string): Date | null => {
  try {
    if (date && date.length === 10) {
      const day: number = parseInt(date.substring(0, 2));
      const month: number = parseInt(date.substring(3, 5));
      const year: number = parseInt(date.substring(6, 10));
      const formattedDate: Date = new Date(Date.UTC(year, month - 1, day));
      return formattedDate;
    }
  } catch (error) {
    console.log(error);
  }
  return null;
};

export const monthsForLocale = (
  localeName = 'no',
  monthFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { month: monthFormat })
    .format;
  const months: string[] = [];
  for (let month = 0; month < 12; month++) {
    months.push(format(new Date(Date.UTC(2020, month))));
  }
  return months;
};

export const weekdaysForLocale = (
  localeName = 'no',
  dayFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { weekday: dayFormat })
    .format;
  const weekdays: string[] = [];
  for (let day = 0; day < 7; day++) {
    weekdays.push(format(new Date(Date.UTC(2020, 7, day + 2)))); // add 2 days to let the week start on Sunday
  }
  return weekdays;
};
