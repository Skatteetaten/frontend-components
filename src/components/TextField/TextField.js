import React, { isValidElement } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { Label } from 'office-ui-fabric-react/lib-commonjs/Label';
import { IconButton } from 'office-ui-fabric-react/lib-commonjs/Button';
//import { Callout, DirectionalHint } from 'office-ui-fabric-react/lib-commonjs/Callout';
import Callout from '../Callout/Callout';
import {
  TextField as FabricTextField,
  MaskedTextField as FabricMaskedTextField
} from 'office-ui-fabric-react/lib-commonjs/TextField';
import { getClassNames } from './TextField.classNames';
import { getClassNames as getLabelClassNames } from './TextFieldLabel.classNames';
import { isUndefined } from 'util';
import ErrorMessage from '../ErrorMessage/ErrorMessage';

class MaskedTextField extends FabricMaskedTextField {
  constructor(props) {
    super(props);
    this._skipComponentRefResolution = false;
  }
}
/**
 * @visibleName TextField (Tekstfelt)
 */
export default class TextField extends React.PureComponent {
  static propTypes = {
    /** Beskrivende informasjon til skjermlesere */
    ariaLabel: PropTypes.string,
    /** Automatisk utvide høyde (ved multiline) */
    autoAdjustHeight: PropTypes.bool,
    /** Benyttes når teksten for et readonly tekstfelt skal fremheves  */
    boldText: PropTypes.bool,
    /** Fjerne rammen rundt feltet */
    borderless: PropTypes.bool,
    /** Bestemmer om hjelptekst/varseltekst skal legge seg mellom label og tekstfelt eller flytende over innhold */
    calloutFloating: PropTypes.bool,
    /** For ytterligere stiling */
    className: PropTypes.string,
    /** Bestemmer om ett readonly felt skal være alltid redigerbart om det er tomt */
    editableWhenEmpty: PropTypes.bool,
    /** Benyttes når et readonly felt skal være redigertbart  */
    editable: PropTypes.bool,
    /** Tilhørende feilmelding */
    errorMessage: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
    /** Tilhørende hjelpetekst */
    help: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
    /** Størrelse på tekstfelt som skal benyttes */
    inputSize: PropTypes.oneOf(['normal', 'large']),
    /** Emnet over feltet */
    label: PropTypes.string,
    /** Størrelse på label */
    labelSize: PropTypes.oneOf(['small', 'large']),
    /** Felt med flere linjer */
    multiline: PropTypes.bool,
    /** Callback når input endres  */
    onChange: PropTypes.func,
    /** Tekst inni feltet som vises før man skriver */
    placeholder: PropTypes.string,
    /** Tekst inni feltet som alltid legges til foran, f.eks. www */
    prefix: PropTypes.string,
    /** Om feltet er inaktivt */
    readonly: PropTypes.bool,
    /** Tekst inni feltet som alltid legges til bak, f.eks. kr */
    suffix: PropTypes.string,
    /** Label på linje med verdi */
    underlined: PropTypes.bool,
    /** Tilhørende varseltekst */
    warning: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
    /** Formateringstreng som definerer oppførsel. Backslash vil escape et tegn. Spesialformat er:  '9': [0-9] 'a': [a-zA-Z] '*': [a-zA-Z0-9]. Eksempel tlf.nr 999 99 999 */
    mask: PropTypes.string,
    /** Antall rader som skal vises i feltet når multiline er satt */
    rows: PropTypes.number
  };

  static defaultProps = {
    inputSize: 'normal',
    editableWhenEmpty: false,
    labelSize: 'small',
    boldText: true,
    calloutFloating: false
  };

  constructor(props) {
    super(props);
    this.state = {
      isCalloutVisible: false,
      editMode: this.editMode(),
      value: '',
      activeCallout: '',
      lineBreak: false
    };
  }

  render() {
    const {
      children,
      onRenderLabel,
      readonly,
      disabled,
      inputSize,
      boldText,
      className,
      mask,
      ...props
    } = this.props;

    let TextFieldType = FabricTextField;
    let value = props.value;
    if (mask) {
      TextFieldType = MaskedTextField;
      value = value ? value : '';
    }

    const setValue = () => {
      if (this.props.suffix && this.props.readonly && !this.state.editMode) {
        return value + ' ' + this.props.suffix;
      }
      return value;
    };
    const { errorMessage, ...filteredProps } = this.props;

    return (
      <div className={classnames(getClassNames(this.props), className)}>
        <TextFieldType
          {...filteredProps}
          value={setValue()}
          ariaLabel={props.label}
          disabled={disabled}
          readOnly={this.state.editMode ? '' : readonly}
          className={classnames(
            getClassNames({ ...this.props, editMode: this.state.editMode }),
            className
          )}
          onRenderLabel={onRenderLabel || this._onRenderLabel}
          onBlur={this._onBlur}
          componentRef={inputField => {
            this.props.componentRef && this.props.componentRef(inputField);
            this._textFieldElement = inputField;
          }}
          inputSize={inputSize}
        >
          {children}
        </TextFieldType>
        {errorMessage && <ErrorMessage>{errorMessage}</ErrorMessage>}
      </div>
    );
  }

  _onRenderLabel = props => {
    const {
      label,
      help,
      warning,
      componentId,
      readOnly,
      editable,
      calloutFloating,
      inputSize
    } = props;
    const styles = getLabelClassNames(props);
    const inputSizeLarge = inputSize === 'large';

    let { isCalloutVisible, activeCallout } = this.state;
    const helpElement = isValidElement(help) ? help : <p>{help}</p>;

    let warningElement = isValidElement(warning) ? warning : <p>{warning}</p>;

    return (
      <div className={styles.labelArea}>
        <span className={styles.label}>
          {label ? <Label htmlFor={componentId}>{label}</Label> : null}
        </span>

        {readOnly && (
          <span className={styles.labelIconArea}>
            {editable && (
              <IconButton
                iconProps={{ iconName: 'Edit' }}
                title="Rediger"
                ariaLabel={'Rediger feltet ' + props.label}
                onClick={this._onEdit}
                className={styles.icon}
              />
            )}
          </span>
        )}

        {help && (
          <span
            className={styles.labelIconArea}
            ref={helpButton => (this._iconButtonElement = helpButton)}
          >
            <IconButton
              iconProps={{ iconName: 'HelpOutline' }}
              title="Hjelp"
              ariaLabel={'Hjelpetekst til feltet ' + props.label}
              onClick={() => this._onClick('helpCallout')}
              className={styles.icon}
            />
          </span>
        )}

        {warning && (
          <span
            className={styles.labelIconArea}
            ref={warningButton => (this._iconButtonElement = warningButton)}
          >
            <IconButton
              iconProps={{ iconName: 'WarningOutline' }}
              title="Varsel"
              ariaLabel={'Varseltekst til feltet ' + props.label}
              onClick={() => this._onClick('warningCallout')}
              className={styles.icon}
            />
          </span>
        )}

        {isCalloutVisible && (
          <Callout
            directionalHint={
              (inputSizeLarge && isUndefined(calloutFloating)) ||
              (!isUndefined(calloutFloating) && !calloutFloating)
                ? Callout.POS_BOTTOM_LEFT
                : Callout.POS_TOP_LEFT
            }
            color={
              activeCallout === 'helpCallout' ? Callout.HELP : Callout.WARNING
            }
            ariaLabel={
              activeCallout === 'helpCallout' ? 'Hjelpetekst' : 'Varseltekst'
            }
            target={this._iconButtonElement}
            onClose={this._onDismiss}
          >
            {activeCallout === 'helpCallout' ? helpElement : warningElement}
          </Callout>
        )}
      </div>
    );
  };

  _onClick = type => {
    let { isCalloutVisible, activeCallout } = this.state;

    if (activeCallout !== type && isCalloutVisible) {
      this.setState({
        activeCallout: type
      });
    } else {
      this.setState({
        activeCallout: type,
        isCalloutVisible: !isCalloutVisible
      });
    }
  };

  _onDismiss = () => {
    this.setState({
      isCalloutVisible: false
    });
  };

  _onEdit = () => {
    this._textFieldElement.focus();
    this.setState({ editMode: true });
  };

  _onBlur = e => {
    this.props.onBlur && this.props.onBlur(e);
    if (this.state.editMode === true) {
      this.setState({
        editMode: this.editMode()
      });
    }
  };

  editMode() {
    if (this.props.editableWhenEmpty && !this.props.value) {
      return this.props.value === '' || this.props.length === 0;
    }
    return false;
  }
}
