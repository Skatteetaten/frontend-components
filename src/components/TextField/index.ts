export { TextField as default } from './TextField';
export * from './TextField';
