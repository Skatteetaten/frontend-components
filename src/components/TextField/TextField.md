** Tekstfelt benyttes når vi ønsker at brukeren skrive inn informasjon av en bestemt type - som oftest i et skjema **

Standard tekstfelt:

```js
import TextField from '@skatteetaten/frontend-components/TextField';

<div style={{ width: '300px' }}>
  <TextField
    id={'my-input-1'}
    label={'Navn'}
    placeholder={'For- og etternavn'}
    value={state.value1}
    onChange={(e, value) => setState({ value1: value })}
  />
</div>;
```

Stor variant av tekstfelt:

```js
import TextField from '@skatteetaten/frontend-components/TextField';

<div style={{ width: '300px' }}>
  <TextField
    id={'my-input-1'}
    label={'Navn'}
    inputSize={'large'}
    placeholder={'For- og etternavn'}
    value={state.value1}
    onChange={(e, value) => setState({ value1: value })}
  />
</div>;
```

Hjelpetekster og advarsel:

```js
import TextField from '@skatteetaten/frontend-components/TextField';

<>
  <div style={{ width: '300px' }}>
    <TextField
      id={'my-helpfield-1'}
      box
      withLeadingIcon="search"
      label="Fullt navn"
      calloutFloating={false}
      placeholder={''}
      value={state.value4}
      onChange={(e, value) => setState({ value4: value })}
      help="Vi trenger å vite navnet ditt dersom vi skal kontakte deg senere."
      calloutFloating={false}
    />
  </div>
  <br />
  <div style={{ width: '150px' }}>
    <TextField
      id={'my-helpfield-2'}
      box
      withLeadingIcon="search"
      label="Antall barn"
      placeholder={''}
      value={'23'}
      calloutFloating={false}
      warning="Er du sikker på at antall barn er riktig?"
      calloutFloating={false}
    />
  </div>
</>;
```

Feilmelding vises i umiddelbar nærhet til feltet:

```js
import TextField from '@skatteetaten/frontend-components/TextField';

<div style={{ width: '160px' }}>
  <TextField
    id={'my-errorfield-2'}
    box
    withLeadingIcon="search"
    label="Inntektsår"
    value={'0218'}
    errorMessage="Inntekståret må være etter 2008."
  />
</div>;
```

Av og til ønsker man å veksle mellom skrive og lesemodus. Da vises det som ren tekst med et blyant-ikon. Når man klikker, blir det skrivefelt:

```js
import TextField from '@skatteetaten/frontend-components/TextField';

const initialState = {
  value: 'Siri Saksbehandler',
  vekt: '4'
};

<div style={{ width: '300px' }}>
  <TextField
    id={'my-readonlyfield'}
    readOnly
    editable
    box
    label="Saksbehandler"
    value={state.value}
    onChange={(e, value) => setState({ value: value })}
    boldText={true}
  />
  <p>Med suffix:</p>
  <TextField
    id={'vektVare'}
    readOnly
    editable
    box
    label="Vekt på vare"
    value={state.vekt}
    onChange={(e, value) => setState({ vekt: value })}
    boldText={true}
    suffix={'kg'}
  />
</div>;
```

Bestemt formatering:

```js
import TextField from '@skatteetaten/frontend-components/TextField';

<div style={{ width: '300px' }}>
  <TextField
    id={'my-input-1'}
    onChange={(e, val) => console.log(val)}
    label={'Org.nummer'}
    placeholder={'999 999 999'}
    value={''}
    mask={'999 999 999'}
    maskChar={''}
  />
</div>;
```

Multiline tekstfelt:

```js
import TextField from '@skatteetaten/frontend-components/TextField';

<div style={{ width: '300px' }}>
  <TextField
    label="Endres til multiline dersom teksten er lengre enn 50 tegn"
    multiline={state.lineBreak}
    onChange={(e, value) => {
      const lineBreak = value.length > 50;
      if (lineBreak !== state.lineBreak) {
        setState({
          lineBreak: lineBreak
        });
      }
    }}
  />
  <br />
  <TextField
    label="Antall rader som vises i multiline modus kan settes"
    multiline
    rows={7}
  />
</div>;
```

```js noeditor
import Accordion from '@skatteetaten/frontend-components/Accordion';
import AccordionItem from '@skatteetaten/frontend-components/Accordion/AccordionItem';

<Accordion>
  <AccordionItem
    isOpen
    toggleContent
    toggleButtonText={'Bruk'}
    stepId={'step-1-1'}
  >
    <p>
      I Designsystemet finnes det mange ulike varisjoner av tekstfelt. Vi har
      med og uten hjelpetekst, feilmeldinger eller lese- og skrivemodus. Vi har
      også mulighet til å bestemme formatering, for eksempel ved
      organiasasjonsnummer.
    </p>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Universell utforming'}
    stepId={'step-1-2'}
  >
    <p>
      Det skal alltid være en programatisk kobling mellom tekstfelt og label,
      slik at skjermleser kan lese opp beskrivelsen av feltet når det er i
      fokus.
    </p>
    <p>
      I publikumsløsninger viser i advarsel- og hjelpetekst mellom label og
      verdi (sett calloutFloating til false). Hjelpetekster i Callout fanges
      ikke like enkelt opp av skjermlesere.
    </p>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Tekniske detaljer'}
    stepId={'step-1-3'}
  >
    <p>
      <a
        href="https://developer.microsoft.com/en-us/fabric#/components/textfield#Implementation"
        target="_blank"
      >
        Se flere tilgjengelige props i Fabric dokumentasjonen
      </a>
    </p>
  </AccordionItem>
</Accordion>;
```
