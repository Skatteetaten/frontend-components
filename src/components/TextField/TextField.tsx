import * as React from 'react';
import Callout from '../Callout/Callout';
import classnames from 'classnames';
import ErrorMessage from '../ErrorMessage/ErrorMessage';
import {
  ITextFieldProps,
  MaskedTextField as FabricMaskedTextField,
  TextField as FabricTextField
} from 'office-ui-fabric-react/lib-commonjs/TextField';
import { getClassNames as getLabelClassNames } from './TextFieldLabel.classNames';
import { getClassNames } from './TextField.classNames';
import { IconButton } from 'office-ui-fabric-react/lib-commonjs/Button';
import { isUndefined } from 'util';
import { Label } from 'office-ui-fabric-react/lib-commonjs/Label';

class MaskedTextField extends FabricMaskedTextField {
  // @ts-ignore TODO
  constructor(props) {
    super(props);
    this._skipComponentRefResolution = false;
  }
}
interface TextFieldProps extends ITextFieldProps {
  /** Benyttes når teksten for et readOnly tekstfelt skal fremheves  */
  boldText?: boolean;
  /** Bestemmer om hjelptekst/varseltekst skal legge seg mellom label og tekstfelt eller flytende over innhold */
  calloutFloating?: boolean;
  /** Bestemmer om ett readOnly felt skal være alltid redigerbart om det er tomt */
  editableWhenEmpty?: boolean;
  /** Benyttes når et readOnly felt skal være redigertbart  */
  editable?: boolean;
  /** Tilhørende hjelpetekst */
  help?: JSX.Element | string;
  /** Størrelse på tekstfelt som skal benyttes */
  inputSize?: 'normal' | 'large';
  /** Størrelse på label */
  labelSize?: 'small' | 'large';
  /** Tekst inni feltet som vises før man skriver */
  placeholder?: string;
  /** Tilhørende varseltekst */
  warning?: JSX.Element | string;
  /** Antall rader som skal vises i feltet når multiline er satt */
  rows?: number;
}
interface TextFieldState {
  isCalloutVisible: boolean;
  editMode: boolean;
  value: string;
  activeCallout: string;
  lineBreak: boolean;
}
/**
 * @visibleName TextField (Tekstfelt)
 */
export default class TextField extends React.PureComponent<
  TextFieldProps,
  TextFieldState
> {
  static defaultProps = {
    inputSize: 'normal',
    editableWhenEmpty: false,
    labelSize: 'small',
    boldText: true,
    calloutFloating: false
  };
  // @ts-ignore TODO
  constructor(props) {
    super(props);
    this.state = {
      isCalloutVisible: false,
      editMode: this.editMode(),
      value: '',
      activeCallout: '',
      lineBreak: false
    };
  }

  render() {
    const {
      children,
      onRenderLabel,
      readOnly,
      disabled,
      inputSize,
      boldText,
      className,
      mask,
      ...props
    } = this.props;

    let TextFieldType = FabricTextField;
    let value = props.value;
    if (mask) {
      // @ts-ignore TODO
      TextFieldType = MaskedTextField;
      value = value ? value : '';
    }

    const setValue = () => {
      if (this.props.suffix && this.props.readOnly && !this.state.editMode) {
        return value + ' ' + this.props.suffix;
      }
      return value;
    };
    const { errorMessage, ...filteredProps } = this.props;
    return (
      <div className={classnames(getClassNames(this.props), className)}>
        {/*
  // @ts-ignore  TODO */}
        <TextFieldType
          {...filteredProps}
          value={setValue()}
          ariaLabel={props.label}
          disabled={disabled}
          readOnly={this.state.editMode ? false : readOnly}
          className={classnames(
            getClassNames({ ...this.props, editMode: this.state.editMode }),
            className
          )}
          onRenderLabel={onRenderLabel || this._onRenderLabel}
          onBlur={this._onBlur}
          componentRef={inputField => {
            // @ts-ignore TODO
            this.props.componentRef && this.props.componentRef(inputField);
            // @ts-ignore TODO
            this._textFieldElement = inputField;
          }}
          inputSize={inputSize}
        >
          {children}
        </TextFieldType>
        {errorMessage && <ErrorMessage>{errorMessage}</ErrorMessage>}
      </div>
    );
  }
  // @ts-ignore TODO
  _onRenderLabel = props => {
    const {
      label,
      help,
      warning,
      componentId,
      readOnly,
      editable,
      calloutFloating,
      inputSize
    } = props;
    const styles = getLabelClassNames(props);
    const inputSizeLarge = inputSize === 'large';

    let { isCalloutVisible, activeCallout } = this.state;
    const helpElement = React.isValidElement(help) ? help : <p>{help}</p>;

    let warningElement = React.isValidElement(warning) ? (
      warning
    ) : (
      <p>{warning}</p>
    );

    return (
      <div className={styles.labelArea}>
        <span className={styles.label}>
          {label ? <Label htmlFor={componentId}>{label}</Label> : null}
        </span>

        {readOnly && (
          <span className={styles.labelIconArea}>
            {editable && (
              <IconButton
                iconProps={{ iconName: 'Edit' }}
                title="Rediger"
                ariaLabel={'Rediger feltet ' + props.label}
                onClick={this._onEdit}
                className={styles.icon}
              />
            )}
          </span>
        )}

        {help && (
          <span
            className={styles.labelIconArea}
            // @ts-ignore TODO
            ref={helpButton => (this._iconButtonElement = helpButton)}
          >
            <IconButton
              iconProps={{ iconName: 'HelpOutline' }}
              title="Hjelp"
              ariaLabel={'Hjelpetekst til feltet ' + props.label}
              onClick={() => this._onClick('helpCallout')}
              className={styles.icon}
            />
          </span>
        )}

        {warning && (
          <span
            className={styles.labelIconArea}
            // @ts-ignore TODO
            ref={warningButton => (this._iconButtonElement = warningButton)}
          >
            <IconButton
              iconProps={{ iconName: 'WarningOutline' }}
              title="Varsel"
              ariaLabel={'Varseltekst til feltet ' + props.label}
              onClick={() => this._onClick('warningCallout')}
              className={styles.icon}
            />
          </span>
        )}

        {isCalloutVisible && (
          <Callout
            directionalHint={
              (inputSizeLarge && isUndefined(calloutFloating)) ||
              (!isUndefined(calloutFloating) && !calloutFloating)
                ? Callout.POS_BOTTOM_LEFT
                : Callout.POS_TOP_LEFT
            }
            color={
              activeCallout === 'helpCallout' ? Callout.HELP : Callout.WARNING
            }
            ariaLabel={
              activeCallout === 'helpCallout' ? 'Hjelpetekst' : 'Varseltekst'
            }
            // @ts-ignore TODO
            target={this._iconButtonElement}
            onClose={this._onDismiss}
          >
            {activeCallout === 'helpCallout' ? helpElement : warningElement}
          </Callout>
        )}
      </div>
    );
  };
  // @ts-ignore TODO
  _onClick = type => {
    let { isCalloutVisible, activeCallout } = this.state;

    if (activeCallout !== type && isCalloutVisible) {
      this.setState({
        activeCallout: type
      });
    } else {
      this.setState({
        activeCallout: type,
        isCalloutVisible: !isCalloutVisible
      });
    }
  };

  _onDismiss = () => {
    this.setState({
      isCalloutVisible: false
    });
  };

  _onEdit = () => {
    // @ts-ignore TODO
    this._textFieldElement.focus();
    this.setState({ editMode: true });
  };
  // @ts-ignore TODO
  _onBlur = e => {
    this.props.onBlur && this.props.onBlur(e);
    if (this.state.editMode === true) {
      this.setState({
        editMode: this.editMode()
      });
    }
  };

  editMode() {
    if (this.props.editableWhenEmpty && !this.props.value) {
      // @ts-ignore TODO
      return this.props.value === '' || this.props.length === 0;
    }
    return false;
  }
}
