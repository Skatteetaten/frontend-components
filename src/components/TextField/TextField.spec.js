import React from 'react';
import { shallow, mount } from 'enzyme';
import { matches } from './../utils/test-utils';
import TextField from './TextField';

function oppsettShallow(props) {
  return shallow(<TextField {...props} />);
}

function oppsettFullDOM(props) {
  return mount(<TextField {...props} />);
}

describe('TextField komponent', () => {
  it('matcher snapshot', () => {
    matches(<TextField />);
  });

  it('rendrer TextFiled med default props', () => {
    const wrapper = oppsettShallow({});

    var input = wrapper.find('StyledTextFieldBase');
    expect(input.prop('inputSize')).toEqual('normal');
    expect(input.prop('editableWhenEmpty')).toEqual(false);
  });

  it('rendrer TextFiled med riktig props', () => {
    const wrapper = oppsettFullDOM({
      ariaLabel: 'textfield-ariaLabel',
      boldText: true,
      borderless: true,
      placeholder: 'textfield-placeholder',
      inputSize: 'large',
      className: 'textfield-classname'
    });
    const textfield = wrapper.find('.ms-TextField');
    const input = wrapper.find('input');

    expect(textfield.prop('className')).toContain('ms-TextField--borderless');
    expect(textfield.prop('className')).toContain('textfield-classname');
    expect(input.prop('placeholder')).toContain('textfield-placeholder');
    expect(wrapper.prop('boldText')).toEqual(true);
    expect(wrapper.prop('ariaLabel')).toEqual('textfield-ariaLabel');
    expect(wrapper.prop('inputSize')).toEqual('large');
  });

  it('rendrer TextFiled label', () => {
    const wrapper = oppsettFullDOM({
      label: 'textfied-label'
    });
    const label = wrapper.find('LabelBase');

    expect(label).toHaveLength(1);
    expect(label.html()).toContain('textfied-label');
  });

  it('rendrer TextFiled med hjelpetekst', () => {
    const wrapper = oppsettFullDOM({
      label: 'Fullt navn',
      help: 'Vi trenger å vite navnet ditt dersom vi skal kontakte deg senere'
    });
    const helpButton = wrapper.find('.ms-Icon');
    expect(helpButton.prop('data-icon-name')).toEqual('HelpOutline');

    helpButton.simulate('click');
    const callout = wrapper.find('StyledCalloutContentBase');
    expect(callout).toHaveLength(1);
    expect(callout.prop('color')).toEqual('lightGreen');
    expect(callout.prop('ariaLabel')).toEqual('Hjelpetekst');
    expect(callout.find('Popup').html()).toContain(
      'Vi trenger å vite navnet ditt dersom vi skal kontakte deg senere'
    );
  });

  it('rendrer TextFiled med hjelpetekst med markup', () => {
    const wrapper = oppsettFullDOM({
      label: 'Fullt navn',
      help: (
        <p>
          Vi trenger å vite navnet <em>ditt</em> dersom vi skal kontakte deg
          senere
        </p>
      )
    });
    const helpButton = wrapper.find('.ms-Icon');

    helpButton.simulate('click');
    const callout = wrapper.find('StyledCalloutContentBase');
    expect(callout).toHaveLength(1);
    expect(
      callout
        .find('Popup')
        .find('em')
        .html()
    ).toContain('ditt');
  });

  it('rendrer TextFiled med varseltekst', () => {
    const wrapper = oppsettFullDOM({
      label: 'Antall barn',
      warning: 'Er du sikker på at antall barn er riktig?'
    });
    const warningButton = wrapper.find('.ms-Icon');

    expect(warningButton.prop('data-icon-name')).toEqual('WarningOutline');

    warningButton.simulate('click');
    const callout = wrapper.find('StyledCalloutContentBase');
    expect(callout).toHaveLength(1);
    expect(callout.prop('color')).toEqual('beige');
    expect(callout.prop('ariaLabel')).toEqual('Varseltekst');
    expect(callout.find('Popup').html()).toContain(
      'Er du sikker på at antall barn er riktig?'
    );
  });

  it('rendrer TextFiled med varseltekst og markup', () => {
    const wrapper = oppsettFullDOM({
      label: 'Antall barn',
      warning: (
        <p>
          Er du sikker på at antall <strong>barn</strong> er riktig?'
        </p>
      )
    });
    const warningButton = wrapper.find('.ms-Icon');

    expect(warningButton.prop('data-icon-name')).toEqual('WarningOutline');

    warningButton.simulate('click');
    const callout = wrapper.find('StyledCalloutContentBase');
    expect(callout).toHaveLength(1);

    expect(
      callout
        .find('Popup')
        .find('strong')
        .html()
    ).toContain('barn');
  });

  it('rendrer TextFiled i lesemodus', () => {
    const wrapper = oppsettFullDOM({
      readOnly: true,
      editable: true
    });
    const editButton = wrapper.find('button.ms-Button--icon');
    let input = wrapper.find('input');

    expect(editButton).toHaveLength(1);
    const icon = editButton.find('.ms-Icon');
    expect(icon.prop('data-icon-name')).toEqual('Edit');
    expect(input.prop('readOnly')).toEqual(true);
    expect(wrapper.state('editMode')).toEqual(false);

    editButton.simulate('click');
    input = wrapper.find('input');
    expect(wrapper.state('editMode')).toEqual(true);
    expect(input.prop('readOnly')).toEqual(false);

    input.simulate('blur');
    input = wrapper.find('input');
    expect(input.prop('readOnly')).toEqual(true);
    expect(wrapper.state('editMode')).toEqual(false);
  });

  it('rendrer TextFiled med feilmelding', () => {
    const wrapper = oppsettFullDOM({
      errorMessage: 'Inntekståret må være etter 2008.'
    });

    expect(wrapper.html()).toContain('role="alert"');
  });

  it('rendrer TextFiled med prefix', () => {
    const wrapper = oppsettFullDOM({
      prefix: 'http://'
    });

    const prefix = wrapper.find('.ms-TextField-prefix');
    expect(prefix).toHaveLength(1);
    expect(prefix.html()).toContain('http://');
  });

  it('rendrer TextFiled med suffix', () => {
    const wrapper = oppsettFullDOM({
      suffix: '.com'
    });

    const suffix = wrapper.find('.ms-TextField-suffix');
    expect(suffix).toHaveLength(1);
    expect(suffix.html()).toContain('.com');
  });

  it('rendrer TextField med flere linjer', () => {
    const wrapper = oppsettFullDOM({
      multiline: true
    });

    expect(wrapper.find('textarea')).toHaveLength(1);
  });

  it('rendrer TextField med flere linjer med satt antall linjer som vises', () => {
    const wrapper = oppsettFullDOM({
      multiline: true,
      rows: 7
    });

    expect(wrapper.find('textarea').prop('rows')).toEqual(7);
  });

  it('skal kjøre on onChange når input feltet oppdateres', () => {
    const onChangeMock = jest.fn();
    const wrapper = mount(<TextField onChange={() => onChangeMock()} />);
    const input = wrapper.find('input');

    expect(onChangeMock.mock.calls.length).toBe(0);
    input.simulate('change', { target: { value: 'abc' } });
    expect(onChangeMock.mock.calls.length).toBe(1);
  });
});

describe('MaskedTextField', () => {
  it('Virker med streng-verdi', () => {
    const value = 'AAAbbbCCC';
    const wrapper = mount(<TextField mask={'aaa aaa aaa'} value={value} />);

    expect(wrapper.prop('value')).toEqual(value);
  });

  it('Virker med tall-verdi', () => {
    const value = 123456789;
    const wrapper = mount(<TextField mask={'999 999 999'} value={value} />);

    expect(wrapper.prop('value')).toEqual(value);
  });

  it('Virker med undefined-verdi', () => {
    const value = undefined;
    const wrapper = mount(<TextField mask={'999 999 999'} value={value} />);

    expect(wrapper.prop('value')).toEqual(value);
  });

  it('Virker med tom streng', () => {
    const value = '';
    const wrapper = mount(<TextField mask={'999 999 999'} value={value} />);

    expect(wrapper.prop('value')).toEqual(value);
  });

  it('Virker med null-verdi', () => {
    const value = null;
    const wrapper = mount(<TextField mask={'999 999 999'} value={value} />);

    expect(wrapper.prop('value')).toEqual(value);
  });
});
