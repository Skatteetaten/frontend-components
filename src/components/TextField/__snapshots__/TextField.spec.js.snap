// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField komponent matcher snapshot 1`] = `
<div
  className=

      & .ms-TextField-field {
        font-size: 1rem;
      }
      && .ms-TextField-fieldGroup {
        border-color: #1d1d1d;
        border-radius: 0px;
      }
      && .ms-TextField-fieldGroup:focus {
        border: #a82e39;
      }
      &.is-active .ms-TextField-fieldGroup {
        border: 1px solid #1362ae;
      }
      &.ms-TextField--underlined .ms-TextField-wrapper {
        border: 1px solid #1d1d1d;
      }
      &.ms-TextField--underlined .ms-TextField-wrapper:hover {
        border: 1px solid rgba(0, 0, 0, 0.95);
      }
      &.is-active.ms-TextField--underlined .ms-TextField-wrapper {
        border: 1px solid #1d1d1d;
        outline: 1px solid #1d1d1d;
      }
      & .ms-TextField-field[readOnly] {
        background: transparent;
        border: none;
        cursor: default;
        font-weight: 400;
        outline: none;
        padding-left: 1px;
      }
      & .ms-TextField-field[readOnly]:focus {
        font-weight: 400;
      }
      & .ms-TextField-field[disabled] {
        color: #666666;
      }
      & .ms-Button-icon {
        font-size: 18px;
      }
      & .ms-TextField-errorMessage {
        color: #a82e39;
        font-size: 0.875rem;
        font-weight: 500;
        padding-left: 20px;
        position: relative;
      }
      & .ms-TextField-errorMessage::before {
        content: 'ï€¨';
        display: block;
        font-family: "Material Icons";
        font-size: 16px;
        left: 0px;
        margin-right: 3px;
        position: absolute;
        top: 6px;
      }
      & textarea.ms-TextField-field {
        resize: none;
      }
      &.is-disabled .ms-TextField-fieldGroup {
        background-color: #e8e8e8;
        border-color: #bbbbbb;
        border-style: solid;
        border-width: 1px;
      }
      &.is-disabled .ms-TextField-field {
        color: #1d1d1d;
      }
      &.is-disabled .ms-TextField-field:hover {
        cursor: not-allowed;
      }
  id="textfield-123"
>
  <div
    className="calloutLabelWrapper-43"
  />
  <div
    className="ms-TextField SkeTextField-51"
  >
    <div
      className="ms-TextField-wrapper"
    >
      <div
        className="ms-TextField-fieldGroup fieldGroup-52"
      >
        <input
          aria-invalid={false}
          className="ms-TextField-field field-53"
          id="textfield-123-input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
